pkg/controller/datavolume/snapshot-clone-controller.go

```diff
@@ -0,0 +1,486 @@
+	if dataVolume.Spec.Source.Snapshot == nil {
+		return errors.Errorf("no source set for clone datavolume")
+	}
+	sourceNamespace := dataVolume.Spec.Source.Snapshot.Namespace
+	if sourceNamespace == "" {
+		sourceNamespace = dataVolume.Namespace
+	}
+	token, ok := dataVolume.Annotations[cc.AnnCloneToken]
+	if !ok {
+		return errors.Errorf("no clone token")
+	}
+	pvc.Annotations[cc.AnnCloneToken] = token
+	pvc.Annotations[cc.AnnCloneRequest] = sourceNamespace + "/" + dataVolume.Spec.Source.Snapshot.Name
+	return nil
+}
+
+func (r SnapshotCloneReconciler) sync(log logr.Logger, req reconcile.Request) (dataVolumeCloneSyncResult, error) {
+	syncRes, syncErr := r.syncSnapshotClone(log, req)
+	if err := r.syncUpdate(log, &syncRes.dataVolumeSyncResult); err != nil {
+		syncErr = err
+	}
+	return syncRes, syncErr
+}
+
+func (r SnapshotCloneReconciler) syncSnapshotClone(log logr.Logger, req reconcile.Request) (dataVolumeCloneSyncResult, error) {
+	res, syncErr := r.syncCommon(log, req, r.cleanup, r.prepare)
+	syncRes := dataVolumeCloneSyncResult{dataVolumeSyncResult: *res}
+	if syncErr != nil || syncRes.result != nil {
+		return syncRes, syncErr
+	}
+
+	pvc := syncRes.pvc
+	pvcSpec := syncRes.pvcSpec
+	datavolume := syncRes.dvMutated
+	transferName := getTransferName(datavolume)
+
+	pvcPopulated := pvcIsPopulated(pvc, datavolume)
+	_, prePopulated := datavolume.Annotations[cc.AnnPrePopulated]
+
+	if pvcPopulated || prePopulated {
+		return syncRes, nil
+	}
+
+	nn := types.NamespacedName{Namespace: datavolume.Spec.Source.Snapshot.Namespace, Name: datavolume.Spec.Source.Snapshot.Name}
+	snapshot := &snapshotv1.VolumeSnapshot{}
+	if err := r.client.Get(context.TODO(), nn, snapshot); err != nil {
+		return syncRes, err
+	}
+	if snapshot.Status == nil || snapshot.Status.ReadyToUse == nil || !*snapshot.Status.ReadyToUse {
+		return syncRes, fmt.Errorf("Snapshot %s/%s not ready to use, can't clone", snapshot.Name, snapshot.Namespace)
+	}
+
+	fallBackToHostAssisted, err := r.evaluateFallBackToHostAssistedNeeded(datavolume, pvcSpec, snapshot)
+	if err != nil {
+		return syncRes, err
+	}
+
+	if pvc == nil {
+		if !fallBackToHostAssisted {
+			res, err := r.reconcileRestoreSnapshot(log, datavolume, snapshot, pvcSpec, transferName, &syncRes)
+			syncRes.result = &res
+			return syncRes, err
+		}
+
+		if err := r.createTempHostAssistedSourcePvc(datavolume, snapshot, pvcSpec, &syncRes); err != nil {
+			return syncRes, err
+		}
+		targetHostAssistedPvc, err := r.createPvcForDatavolume(datavolume, pvcSpec, r.updateAnnotations)
+		if err != nil {
+			if cc.ErrQuotaExceeded(err) {
+				r.syncDataVolumeStatusPhaseWithEvent(&syncRes, cdiv1.Pending, nil,
+					Event{
+						eventType: corev1.EventTypeWarning,
+						reason:    cc.ErrExceededQuota,
+						message:   err.Error(),
+					})
+			}
+			return syncRes, err
+		}
+		pvc = targetHostAssistedPvc
+	}
+
+	if fallBackToHostAssisted {
+		if err := r.ensureExtendedToken(pvc); err != nil {
```

Comment:
Great point! I see it's really possible for readyToUse to take a while:
https://github.com/kubevirt/kubevirt/pull/8640

Thing is, we need a target PVC to exist for `ensureExtendedToken`.
Should I drop waiting for readyToUse, since everything will converge eventually when it becomes ready?

Reply:
yeah I guess we don't have to wait but should probably make sure that the snapshot does not have the error set

Reply:
do you mean `snapshot.status.error`? what if some csi driver is notoriously known for setting new errors for a long time, until snapshot is ready?
we would be back to the same problem

Reply:
yeah that's the field.  I think it will be very unlikely for a snapshot to be in error state then be good

Reply:
done